src/main/java/net/minecraftforge/common/capabilities/AttachCapabilitiesEvent.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.common.capabilities;
+
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraftforge.eventbus.api.Event;
+
+/**
+ * The AttachCapabilitiesEvent is used to attach {@link IAttachedCapabilityProvider}s to game objects.<br>
+ * The firing of this event is determined by the provider object, and is not consistent across usages.<br>
+ * For most cases, this is fired whenever the {@link CapabilityDispatcher} must be resolved.<br>
+ * This means that typically, this event is not fired until capabilities are attempted to be accessed for a particular object.
+ */
+public abstract class AttachCapabilitiesEvent<T extends ICapabilityProvider> extends Event
+{
+
+    protected final T obj;
+    protected final Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> providers = new HashMap<>();
+    protected final Map<CapabilityType<?>, IAttachedCapabilityProvider<?, T>> view = Collections.unmodifiableMap(providers);
+    protected final Map<ResourceLocation, IAttachedCapabilityProvider<?, T>> byName = new HashMap<>();
+    protected final Map<ResourceLocation, IAttachedCapabilityProvider<?, T>> byNameView = Collections.unmodifiableMap(byName);
+
+    protected AttachCapabilitiesEvent(T obj)
+    {
+        this.obj = obj;
+    }
+
+    /**
+     * Returns the object that is having capabilities attached.
+     */
+    public final T getObject()
+    {
+        return this.obj;
+    }
+
+    /**
+     * Returns the number of successful attachments.
+     */
+    public final int getSize()
+    {
+        return this.providers.size();
+    }
+
+    /**
+     * Adds a {@link IAttachedCapabilityProvider} to this object.<br>
+     * Only one provider may exist for any given {@link CapabilityType}, on a first-come-first-serve basis.<br>
+     * Additionally, the ID of all providers must be unique across the all {@link CapabilityType}s.
+     * 
+     * @param provider The provider being attached to this game object.
+     * @return true, if this provider was successfully attached, otherwise false.
+     */
+     public boolean addCapability(IAttachedCapabilityProvider<?, T> provider)
+     {
+        CapabilityType<?> type = provider.getType();
+        // Review Note : Potentially throw on duplicate name?  Would indicate major issues.
+        if(this.providers.containsKey(type) || this.byName.containsKey(provider.getId())) return false;
```

Comment:
Imo duplicate CapabilityTypes are fine to just ignore with `return false`, however duplicate names should be a programming error.

Reply:
Yeah I figured that would be the case, I think I'll update that accordingly.  Duplicate names should indicate a hard error.